cli/src/commands/squash.rs

```diff
@@ -227,105 +222,65 @@ from the source will be moved into the destination.
         let selected_tree_id =
             diff_selector.select(&parent_tree, &source_tree, matcher, format_instructions)?;
         let selected_tree = tx.repo().store().get_root_tree(&selected_tree_id)?;
-        let abandon = !keep_emptied && selected_tree.id() == source_tree.id();
-        if !abandon && selected_tree_id == parent_tree.id() {
-            // Nothing selected from this commit. If it's abandoned (i.e. already empty), we
-            // still include it so `jj squash` can be used for abandoning an empty commit in
-            // the middle of a stack.
-            continue;
+
+        let commit = rewrite::CommitToSquash::new(tx.repo(), source.clone(), selected_tree)?;
+        if !keep_emptied && commit.is_full_selection() {
+            abandoned_commits.push(source);
```

Comment:
It's weird that we have to test the same condition by caller and callee. Maybe we can extract description handling as a callback? It might be also good to make `rewrite::squash_commits()` return `CommitBuilder` and let caller do postprocessing as needed.

Reply:
I tried both of these approaches first but couldn't get them to work due to lifetime issues.  `combine_descriptions()` requires an immutable borrow of `tx.base_workspace_helper()` which can't overlap with `CommitBuilder` since it borrows the repo mutably.

Reply:
Ah, that's convenient.  Added a commit, done.